src/gpuhunt/providers/cudocompute.py

```diff
@@ -0,0 +1,458 @@
+import asyncio
+from collections import namedtuple
+from concurrent.futures import ThreadPoolExecutor
+from itertools import chain
+from typing import Optional, List
+from cudo_compute import cudo_api
+from gpuhunt import QueryFilter, RawCatalogItem
+from gpuhunt._internal.constraints import KNOWN_GPUS
+from gpuhunt.providers import AbstractProvider
+import logging
+
+CpuMemoryGpu = namedtuple("CpuMemoryGpu", ["cpu", "memory", "gpu"])
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class CudoComputeProvider(AbstractProvider):
+    NAME = "cudocompute"
+
+    def get(
+            self, query_filter: Optional[QueryFilter] = None, balance_resources: bool = True
+    ) -> List[RawCatalogItem]:
+        offers = asyncio.run(self.fetch_all_vm_types())
+        return sorted(offers, key=lambda i: i.price)
+
+    async def fetch_all_vm_types(self):
+        executor = ThreadPoolExecutor(max_workers=10)
+        tasks = []
+        for cmg in GPU_MACHINES:
+            task = asyncio.create_task(self.fetch_vm_type(cmg.cpu, cmg.memory, cmg.gpu, executor))
+            tasks.append(task)
+        results = await asyncio.gather(*tasks)
+        executor.shutdown(wait=True)
+        return list(chain.from_iterable(results))
+
+    def get_raw_catalog_list(self, vm_machine_type_list, vcpu, memory, gpu):
+        raw_list = []
+        for vm in vm_machine_type_list.host_configs:
+            raw = RawCatalogItem(
+                instance_name=vm.machine_type,
+                location=vm.data_center_id,
+                spot=False,
+                price=round((float(vm.total_price_hr.value) + float(vm.storage_gib_price_hr.value)), 5),
+                cpu=vcpu,
+                memory=memory,
+                gpu_count=gpu,
+                gpu_name=gpu_name(vm.gpu_model),
+                gpu_memory=get_memory(gpu_name(vm.gpu_model)),
+                disk_size=None
+            )
+            raw_list.append(raw)
+        return raw_list
+
+    async def fetch_vm_type(self, vcpu, memory_gib, gpu, executor):
+        loop = asyncio.get_running_loop()
+        result = await loop.run_in_executor(
+            executor,
+            lambda: cudo_api.virtual_machines().list_vm_machine_types(vcpu=vcpu, memory_gib=memory_gib, gpu=gpu)
+        )
+        return self.get_raw_catalog_list(result, vcpu, memory_gib, gpu)
+
+
+def gpu_name(name: str) -> Optional[str]:
+    if not name:
+        return None
+
+    result = GPU_MAP.get(name)
+
+    if result is None:
+        logging.warning("There is no '%s' in GPU_MAP", name)
```

Comment:
Since this code is working in CI/CD, it's better to raise an exception

Reply:
Done